### PR TITLE
Added probe attenutation

### PR DESCRIPTION
**Pull request by dipusone**
Added the probe attenuation settings. The probe attenuation is applied directly on the acquired samples in order to reflect the real volteges even in the math/info sections (EDIT: see TODOS).

- Each channel get a probe and the relative settings
- For each channel/probe is possible to set a list of gains separated by a comma 
- In the voltage dock is possible to select the attenuation for the probe

TODOS:
- Since the data processing has changed from the previous version i have to reimplement the voltage modification.
- Possibly add a class to store the probe settings

### CODE REVIEW

File: openhantek/src/configdialog/DsoConfigProbePage.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
+//
```
Comment (davidgraeff) [inline openhantek/src/configdialog/DsoConfigProbePage.cpp]: I forgot to add this information to the contribution guideline, but please use the license header line

File: openhantek/src/configdialog/DsoConfigProbePage.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
+//
```
Comment (davidgraeff) [inline openhantek/src/configdialog/DsoConfigProbePage.h]: I forgot to add this information to the contribution guideline, but please use the license header line

File: openhantek/src/configdialog/DsoConfigProbePage.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// Created by jacopo on 08/01/18.
+//
+
+#ifndef OPENHANTEKPROJECT_DSOCONFIGPROBEPAGE_H
+#define OPENHANTEKPROJECT_DSOCONFIGPROBEPAGE_H
```
Comment (davidgraeff) [inline openhantek/src/configdialog/DsoConfigProbePage.h]: `#pragma once` is used in this project, no header guards

File: openhantek/src/docks/VoltageDock.cpp [hunk 4]
```diff
@@ -129,3 +155,35 @@ void VoltageDock::setUsed(ChannelID channel, bool used) {
     QSignalBlocker blocker(channelBlocks[channel].usedCheckBox);
     channelBlocks[channel].usedCheckBox->setChecked(used);
 }
+
+int VoltageDock::setProbeGain(ChannelID channel, double probeGain) {
+    if (channel < 0 ||channel >= spec->channels) return -1;
+
+    QSignalBlocker blocker(channelBlocks[channel].probeGainCombobox);
+    int index = (int) (std::find(scope->voltage[channel].probeGainSteps.begin(),
```
Comment (davidgraeff) [inline openhantek/src/docks/VoltageDock.cpp]: This was done before with gain as well, that the gain value is stored and the set-Method needed to use `find` to determine the proper index. I prefer the way that the index is stored instead and if the probe value is required, it is looked up like `available_probes[probe_index]`.

File: openhantek/src/hantekdso/controlsettings.h [hunk 5]
```diff
@@ -51,6 +51,7 @@ struct ControlSettings {
     ~ControlSettings();
     ControlSettingsSamplerate samplerate;        ///< The samplerate settings
     std::vector<ControlSettingsVoltage> voltage; ///< The amplification settings
+    std::vector<double> probeGain; ///< The gain probe gain for the current channel
```
Comment (davidgraeff) [inline openhantek/src/hantekdso/controlsettings.h]: I would prefer a probeIndex/probeId here

File: openhantek/src/hantekdso/hantekdsocontrol.cpp [hunk 6]
```diff
@@ -1223,8 +1233,8 @@ void HantekDsoControl::run() {
             this->_samplingStarted = false;
 
             // Start next capture if necessary by leaving out the break statement
-
             if (!this->sampling) break;
+#if defined(__clang__)
```
Comment (davidgraeff) [inline openhantek/src/hantekdso/hantekdsocontrol.cpp]: Which compiler do you use? This part works with G++ and clang and if another special case is required, we should add it. C++14 added the fallthrough attribute officially.

File: openhantek/src/mainwindow.h [hunk 7]
```diff
@@ -37,6 +37,8 @@ public slots:
     // Central widgets
     DsoWidget *dsoWidget;
 
+    VoltageDock *voltageDock;
```
Comment (davidgraeff) [inline openhantek/src/mainwindow.h]: Why is this necessary?

File: openhantek/src/scopesettings.h [hunk 8]
```diff
@@ -55,14 +55,19 @@ struct DsoSettingsScopeVoltage {
...
-        int rawValue;
-    };
-    QString name;          ///< Name of this channel
-    bool inverted = false; ///< true if the channel is inverted (mirrored on cross-axis)
-    bool used = false;     ///< true if this channel is enabled
+    bool inverted = false;      ///< true if the channel is inverted (mirrored on cross-axis)
+                   ///< Different enums, coupling for real- and mode for math-channels
+    Dso::MathMode math;
+    unsigned couplingIndex = 0;
+    int rawValue;
```
Comment (davidgraeff) [inline openhantek/src/scopesettings.h]: math, couplingIndex and rawValue should be in a union. Only rawvalue is read/written by the settings object.

File: openhantek/src/scopesettings.h [hunk 9]
```diff
@@ -55,14 +55,19 @@ struct DsoSettingsScopeVoltage {
...
+
+    std::vector<double> probeGainSteps; ///< Probe gain steps for channel
+    double probe_gain = 1; ///< Default value for the probe gain
```
Comment (davidgraeff) [inline openhantek/src/scopesettings.h]: Isn't it enough to have `probeStepIndex` and `defaultValues`?

File: openhantek/src/settings.h [hunk 10]
```diff
@@ -21,6 +21,7 @@ class DsoSettings {
 
     DsoSettingsExport exporting; ///< General options of the program
     DsoSettingsScope scope;     ///< All oscilloscope related settings
+    const Dso::ControlSpecification* deviceSpecification; ///< Device specification settings
```
Comment (davidgraeff) [inline openhantek/src/settings.h]: That slipped in I guess? I have removed the `deviceSpecification` again. It is always accessible via `DsoControl->getDevice()->getModel().spec`

File: openhantek/src/utils/printutils.h [hunk 11]
```diff
@@ -9,7 +9,7 @@
 //////////////////////////////////////////////////////////////////////////////
 /// \enum Unit utils/printutils.h
 /// \brief The various units supported by valueToString.
-enum Unit { UNIT_VOLTS, UNIT_DECIBEL, UNIT_SECONDS, UNIT_HERTZ, UNIT_SAMPLES, UNIT_COUNT };
+enum Unit { UNIT_VOLTS, UNIT_DECIBEL, UNIT_SECONDS, UNIT_HERTZ, UNIT_SAMPLES, UNIT_COUNT, UNIT_TIMES};
```
Comment (davidgraeff) [inline openhantek/src/utils/printutils.h]: UNIT_COUNT should always be the last item

File: openhantek/src/configdialog/DsoConfigProbePage.h (hunk 3 above)
Comment (dipusone) [reply openhantek/src/configdialog/DsoConfigProbePage.h]: I'll remove them! Just an habit.

File: openhantek/src/hantekdso/hantekdsocontrol.cpp (hunk 6 above)
Comment (dipusone) [reply openhantek/src/hantekdso/hantekdsocontrol.cpp]: Gcc 4.8.5 and failed to interpret the subsequent __has... so i had to add a check to compile

File: openhantek/src/mainwindow.h (hunk 7 above)
Comment (dipusone) [reply openhantek/src/mainwindow.h]: Because otherwise QT cannot connect the signal from the Qdialog fished to the slot in the VoltageDock.

File: openhantek/src/scopesettings.h (hunk 8 above)
Comment (dipusone) [reply openhantek/src/scopesettings.h]: The union would not compile on gcc 4.8.5 and all the oscilloscope would not work.

File: openhantek/src/scopesettings.h (hunk 9 above)
Comment (dipusone) [reply openhantek/src/scopesettings.h]: This was because this was used before directly on the raw data and i would not have to pay the cost for the array access.

File: openhantek/src/settings.h (hunk 10 above)
Comment (dipusone) [reply openhantek/src/settings.h]: Yep, i couldn't find it. I will remove it!

File: openhantek/src/utils/printutils.h (hunk 11 above)
Comment (dipusone) [reply openhantek/src/utils/printutils.h]: I'modify it.

File: openhantek/src/configdialog/DsoConfigProbePage.cpp (hunk 1 above)
Comment (dipusone) [reply openhantek/src/configdialog/DsoConfigProbePage.cpp]: I'l modify it!

File: openhantek/src/scopesettings.h (hunk 8 above)
Comment (dipusone) [reply openhantek/src/scopesettings.h]: With the union and and compiled with gcc 4.8.5 the oscilloscope goes in SIGSEGV once it tries to access any member of the union.

File: openhantek/src/hantekdso/controlsettings.h (hunk 5 above)
Comment (dipusone) [reply openhantek/src/hantekdso/controlsettings.h]: With probeIndex  you mean the index of the gain in the list of possible gain of a probe?

File: openhantek/src/configdialog/DsoConfigProbePage.cpp [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+/// \brief Saves the new settings.
+void DsoConfigProbePage::saveSettings() {
+    //TODO find a way to refresh a widget
+    for(unsigned int channel = 0; channel < this->settings->scope.voltage.size(); ++channel) {
+        if(channel < spec->channels) {
+            // Clear the list
+            this->settings->scope.voltage[channel].probeGainSteps.clear();
+            QStringList values = this->probeAttenuations[channel]->text().split(',');
+            // Try to convert the values in booleans and save them the the array
```
Comment (davidgraeff) [inline openhantek/src/configdialog/DsoConfigProbePage.cpp]: ...to convert to doubles (not booleans)

File: openhantek/src/configdialog/DsoConfigProbePage.h [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    public:
+    DsoConfigProbePage(DsoSettings *settings, const Dso::ControlSpecification *spec, QWidget *parent = 0);
+    ~DsoConfigProbePage();
+
+    public slots:
+        void saveSettings();
+
+
+    private:
+        QList<QLabel *> probeLabel;
```
Comment (davidgraeff) [inline openhantek/src/configdialog/DsoConfigProbePage.h]: If possible use std::vector (or QVector). QList is a very powerful container class that comes with a big overhead as well

File: openhantek/src/docks/VoltageDock.cpp [hunk 14]
```diff
@@ -129,3 +154,38 @@ void VoltageDock::setUsed(ChannelID channel, bool used) {
     QSignalBlocker blocker(channelBlocks[channel].usedCheckBox);
     channelBlocks[channel].usedCheckBox->setChecked(used);
 }
+
+void VoltageDock::setProbeGain(ChannelID channel, unsigned int probeIndex) {
+    if (channel < 0 ||channel >= spec->channels) return ;
```
Comment (davidgraeff) [inline openhantek/src/docks/VoltageDock.cpp]: ChannelID is a typedef to unsigned (and channels will always be a positive number), therefore you don't need to check for "channel < 0".

File: openhantek/src/docks/VoltageDock.cpp [hunk 15]
```diff
@@ -129,3 +154,38 @@ void VoltageDock::setUsed(ChannelID channel, bool used) {
...
+            if(scope->voltage[channel].probeStepIndex >= scope->voltage[channel].probeGainSteps.size()){
+                this->scope->voltage[channel].probeStepIndex = 0;
+                emit probeGainChanged(channel, 0);
+
+            }
+
+            }
+    }
+
+}
```
Comment (davidgraeff) [inline openhantek/src/docks/VoltageDock.cpp]: Newline missing

File: openhantek/src/docks/VoltageDock.cpp [hunk 16]
```diff
@@ -129,3 +154,38 @@ void VoltageDock::setUsed(ChannelID channel, bool used) {
...
+            QStringList probeGainStrings;
+            for(double probe_gain: scope->voltage[channel].probeGainSteps)
+                    probeGainStrings << valueToString(probe_gain, UNIT_TIMES, 0);
+            channelBlocks[channel].probeGainCombobox->addItems(probeGainStrings);
+            // If the index of the list is outside the new list, set index to 0 and notify others
+            if(scope->voltage[channel].probeStepIndex >= scope->voltage[channel].probeGainSteps.size()){
+                this->scope->voltage[channel].probeStepIndex = 0;
+                emit probeGainChanged(channel, 0);
+
+            }
```
Comment (davidgraeff) [inline openhantek/src/docks/VoltageDock.cpp]: Have you applied the clang-format tool? In QtCreator you can right click the "format" target in the projects tree view and click "execute". Be careful, this will format all the code (and might add format changes to files you haven't touched). You can also execute Clang-Format from the Tools menu.

File: openhantek/src/hantekdso/hantekdsocontrol.cpp [hunk 17]
```diff
@@ -65,6 +65,9 @@ HantekDsoControl::HantekDsoControl(USBDevice *device)
 
     // Apply special requirements by the devices model
     device->getModel()->applyRequirements(this);
+    // Initialize the gain vector with standard and safe values
+    controlsettings.probeGain.resize( specification.channels, 1);
```
Comment (davidgraeff) [inline openhantek/src/hantekdso/hantekdsocontrol.cpp]: Please do this in the controlsettings constructor. But read the next comment first. It might make sense to not apply the probe gain here but in a post processing step.

File: openhantek/src/hantekdso/hantekdsocontrol.cpp [hunk 18]
```diff
@@ -271,14 +275,14 @@ void HantekDsoControl::convertRawDataToSamples(const std::vector<unsigned char>
                     ((unsigned short int)rawData[totalSampleCount + bufferPosition - extraBitsPosition] << shift) &
                     extraBitsMask;
 
-                result.data[channel][pos] = ((double)(low + high) / limit - offset) * gainStep;
+                result.data[channel][pos] = ((double)(low + high) / limit - offset) * gainStep * probeGain;
```
Comment (davidgraeff) [inline openhantek/src/hantekdso/hantekdsocontrol.cpp]: HantekDsoControl should output the pure samples, only adapted to the devices abilities (10bit, limits, offsets). Your signal modification is more a post processing change and belongs into the `post` directory as a separate class that registers to the PostProcessing class. See the math channel generator class in the same directory as an example.

File: openhantek/src/hantekdso/hantekdsocontrol.cpp (hunk 6 above)
Comment (davidgraeff) [reply openhantek/src/hantekdso/hantekdsocontrol.cpp]: Understandable. G++ 4.8.5 is only 2 1/2 years old, that can and should still be supported. Leave it as it is and I will find a solution in a later PR.

File: openhantek/src/mainwindow.cpp [hunk 19]
```diff
@@ -155,6 +156,14 @@ MainWindow::MainWindow(HantekDsoControl *dsoControl, DsoSettings *settings, QWid
         dsoControl->setGain(channel, mSettings->scope.gain(channel) * DIVS_VOLTAGE);
     });
     connect(voltageDock, &VoltageDock::gainChanged, dsoWidget, &DsoWidget::updateVoltageGain);
+
+    connect(voltageDock, &VoltageDock::probeGainChanged, [this, dsoControl, spec, scope](ChannelID channel, unsigned probeGain) {
+        if (channel >= spec->channels) return;
+        // TODO move to the index
+        dsoControl->setProbeGain(channel, scope->voltage[channel].probeGainSteps[scope->voltage[channel].probeStepIndex]);
```
Comment (davidgraeff) [inline openhantek/src/mainwindow.cpp]: If you implement this as a post processing step, that uses settings directly, you don't even need to notify anything because the voltage dock applies the change directly to the settings object. And the block can be removed.

File: openhantek/src/scopesettings.h (hunk 8 above)
Comment (davidgraeff) [reply openhantek/src/scopesettings.h]: Maybe that's a compiler bug. We have a union with an enum class inside. The later is quite new to that compiler version. If you remove the union, math channel and coupling index will never be saved or loaded.

**Comment by davidgraeff** [comment]
I have removed the union. That caused some conflicts now of course.

File: openhantek/src/hantekdso/controlsettings.h (hunk 5 above)
Comment (dipusone) [reply openhantek/src/hantekdso/controlsettings.h]: I would keep this a single gain to avoid the overhead of accessing an array to get the prove gain during the sample acquisition.

File: openhantek/src/configdialog/DsoConfigProbePage.cpp [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+    //Add the labels for each channel
+    for(unsigned int channel = 0; channel < this->settings->scope.voltage.size(); ++channel) {
+        if(channel < spec->channels) {
+            this->probeLabel.append(new QLabel(QApplication::tr("Probe Gain for Channel %L1").arg(channel)));
+            // Fast join
+            QString values;
+            // TODO find a better way
+            if(this->settings->scope.voltage[channel].probeGainSteps.size() > 0) {
+                for (unsigned int idx = 0; idx < this->settings->scope.voltage[channel].probeGainSteps.size(); idx++) {
```
[the remark this answers is not in the document]
Comment (dipusone) [reply openhantek/src/configdialog/DsoConfigProbePage.cpp]: Nice idea, i was quite tired when i wrote this code so may brain couldn't work properly :)

File: openhantek/src/docks/VoltageDock.cpp (hunk 16 above)
Comment (dipusone) [reply openhantek/src/docks/VoltageDock.cpp]: I do use Clion to write my code. Normally I write python and i'n a PEP8 nazi :P. For c++ I didn't know the convention of the project.

File: openhantek/src/hantekdso/hantekdsocontrol.cpp (hunk 18 above)
Comment (dipusone) [reply openhantek/src/hantekdso/hantekdsocontrol.cpp]: My idea was this: the probe attenuation gives in output a voltage that is actually not the real one, but one attenutate, so it would make sense to process the raw voltage to adjust for that.

**Comment by davidgraeff** [comment]
I will close this for now but I'd be happy to accept further contributions.
